scenario/utils.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+import re
+import typing
+from typing import List, Literal, NamedTuple, Optional
+
+if typing.TYPE_CHECKING:
+    from scenario.state import State
+
+
+JUJU_EVT_REGEX = re.compile(r"Emitting Juju event (?P<event>\S+)\.")
+_EVENT_REPR = (
+    r"<(?P<event_cls>\S+) via (?P<charm_name>\S+)/on/(?P<event>\S+)\[(?P<n>\d+)\]>\."
+)
+
+REEMITTING_EVT_REGEX_NEW = re.compile(f"Re-emitting deferred event {_EVENT_REPR}")  # ops >= 2.1
+REEMITTING_EVT_REGEX_OLD = re.compile(f"Re-emitting {_EVENT_REPR}")  # ops < 2.1
+CUSTOM_EVT_REGEX = re.compile(f"Emitting custom event {_EVENT_REPR}")  # ops >= 2.1
+OPERATOR_EVT_REGEX = re.compile(r"Charm called itself via hooks/(?P<event>\S+)\.")
+
+class EventEmissionLog(NamedTuple):
+    name: str
+    source: Literal['juju', 'custom', 'deferral', 'framework']
+    raw: str
+
+
+def match_line(line: str) -> Optional[EventEmissionLog]:
```

Comment:
Wouldn't it be much safer and cleaner to patch ops's `emit` functionality rather than parsing the text of log messages, which are subject to change (as we've seen)?

Reply:
I remember doing that at some point and receiving the feedback:
- why do this when you could obtain the same information from reading the juju-log?
I agree that what you propose is better and exposes more data (the Event instances, for example), I thought we were not all on board with the option. 
If you and @sed-i both think that's viable, I'll refactor to use patching because I like it more as well.

Reply:
Actually, I think the best thing to do is to bake https://github.com/PietroPasotti/harness-extensions/blob/master/libs/capture_events/capture_events.py into Scenario. It would work out of the box, but I reckon it's better to build it in rather than forcing people to use an extension.

On the other hand, it does feel like white-box testing, so perhaps it does not belong to Scenario after all.

Maybe I should write a `scenario-unboxed` library for this kind of stuff :laughing: 


thoughts?

Reply:
> why do this when you could obtain the same information from reading the juju-log?

Was that me and now I'm flip-flopping? If so, sorry. :-) I do think both are hacks, but yeah, the patch approach seems like a much cleaner hack.

Yeah, using capture_events (or copying it in) seems reasonable.